### PR TITLE
build: imgtool integration for MCUboot

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
In https://github.com/zephyrproject-rtos/zephyr/issues/22562#issuecomment-654421847 and following comments, we discuss how to make it easier to build zephyr applications to be chain-loaded by MCUboot. In short, we agreed on auto-generating signed binaries, provided sufficient information about how to generate them is available in .config.

This PR implements that.

If accepted, I will write a follow-up PR that teaches `west flash` to use the signed binaries by default. This will probably involve finally deprecating the `--elf-file`, `--hex-file`, and `--bin-file` runner options and just loading the default locations from `build/zephyr/.config`, with enough wiggle room to override those defaults when MCUboot is in use.

We can leave an escape hatch for user-supplied files using the `--file` option we've discussed adding support for.

The first four patches are just cleanups I noticed along the way and prep work. (@finikorg, I hope you will ACK the rimage changes to support `west sign --quiet`).

The last patch is the main one.

How I am testing, for reference:

```
west build -b nrf52840dk_nrf52840 -s bootloader/mcuboot/boot/zephyr -d build-mcuboot
west flash -d build-mcuboot

west build -b nrf52840dk_nrf52840 -s zephyr/samples/hello_world -- \
   -DCONFIG_BOOTLOADER_MCUBOOT=y \
   -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"/home/mbolivar/zp/bootloader/mcuboot/root-rsa-2048.pem\"

# My board's default flash runner supports --hex-file.
west flash --hex-file build/zephyr/zephyr.signed.hex
```

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
@JordanYates please review. (I can't add you as a reviewer but I would like your feedback.) Thanks.

File: cmake/mcuboot.cmake
Comment (JordanYates) [inline cmake/mcuboot.cmake:75]: Does it makes sense for the output binary to be ` ${ZEPHYR_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.signed.bin`?
Default behaviour would still be `zephyr.bin` and `zephyr.signed.bin`, but `custom_app_name.bin` would generate `custom_app_name.signed.bin`.
Same reasoning would apply to the hex output.

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [inline cmake/mcuboot.cmake:19]: Should this be an error instead?

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [inline cmake/mcuboot.cmake:22]: Should this be an error instead?

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [inline cmake/mcuboot.cmake:30]: Would it make sense to turn this logic upside down? If mcuboot is part of the manifest, the version of imgtool known to work with that revision of mcuboot is the one shipped with it.

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [inline cmake/mcuboot.cmake:56]: Again, should this be an error instead?

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [reply cmake/mcuboot.cmake:75]: I agree.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
@mbolivar-nordic Thank you for working on this. It will greatly reduce the complexity of building signed binaries for most downstream users.

I wonder if the various warnings emitted if west or imgtool.py isn't found although the configuration calls for signing should be errors instead?

**Comment by henrikbrixandersen** [comment]
In our downstream, we always generate both a pre-confirmed and a non-confirmed image when signing. This allows us to use the exact same .elf file for both production flash programming (where we use the pre-confirmed image and merge it with the mcuboot image) and OTA updates (where we use the non-confirmed image and confirm the image externally after upgrade).

Do you think this use-case is generic enough to be supported by the upstream build system? Perhaps with a Kconfig option for providing a pre-confirmed image as well?

File: cmake/mcuboot.cmake
Comment (mbolivar-nordic) [reply cmake/mcuboot.cmake:75]: > Does it makes sense for the output binary to be ${ZEPHYR_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.signed.bin?
> Same reasoning would apply to the hex output.

Agreed.

File: cmake/mcuboot.cmake
Comment (mbolivar-nordic) [reply cmake/mcuboot.cmake:22]: > Should this be an error instead?

@henrikbrixandersen I think that's reasonable. The user has to take explicit action to set MCUBOOT_SIGNATURE_KEY_FILE to a non-default value so the intent is clear here.

**Comment by mbolivar-nordic** [comment]
> Do you think this use-case is generic enough to be supported by the upstream build system? Perhaps with a Kconfig option for providing a pre-confirmed image as well?

Sure, that seems reasonable. Can you suggest some file names?

For example, should zephyr.signed.hex be unconfirmed, and zephyr.signed-confirmed.hex be confirmed?

File: cmake/mcuboot.cmake
Comment (mbolivar-nordic) [reply cmake/mcuboot.cmake:30]: > Would it make sense to turn this logic upside down? If mcuboot is part of the manifest, the version of imgtool known to work with that revision of mcuboot is the one shipped with it.

@henrikbrixandersen My thinking here is that the user has to install imgtool with pip explicitly, because it's not part of any requirements.txt file. So the intent if it's present seems to be to use that one; doing it this way also avoids a subprocess on Windows, where they're more expensive.

If you think this is important, I will reverse the order.

**Comment by mbolivar-nordic** [comment]
I pushed a v2 which changes warnings to errors and respects CONFIG_KERNEL_BIN_NAME.

Open review comments:

- creating confirmed images too, potentially under another Kconfig option (what to name the files in this case?)
- whether to prefer the imgtool that comes with mcuboot or the one installed with pip

File: cmake/mcuboot.cmake
Comment (henrikbrixandersen) [reply cmake/mcuboot.cmake:30]: No, that's a good point. I didn't think of that. Let's keep the order as-is.

**Comment by henrikbrixandersen** [comment]
> Sure, that seems reasonable. Can you suggest some file names?
> 
> For example, should zephyr.signed.hex be unconfirmed, and zephyr.signed-confirmed.hex be confirmed?

@mbolivar-nordic Yes, that sounds like a reasonable naming scheme to me (or zephyr.signed.confirmed.hex?).

**Comment by mbolivar-nordic** [comment]
Pushed a v3:

- Added CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE
- Documentation updates to Kconfig options
- Miscellaneous cleanups

No open review comments.

**Review by henrikbrixandersen — APPROVED** [review]
Thanks, @mbolivar-nordic. Looks good! Only thing missing, I can think of, is a test of this within the tree - but since the path to the key file must be absolute, it's probably hard to get it to work in CI.

**Comment by mbolivar-nordic** [comment]
> Only thing missing, I can think of, is a test of this within the tree - but since the path to the key file must be absolute, it's probably hard to get it to work in CI.

That's not the issue as much as me having no idea how to write the test.

It's easy to write a test to test C or Python behavior, but I don't know of any tests that verify the build artifacts themselves.

I propose that we test this is from the `west flash` test cases once that tool understands how to react to these Kconfig options. Sound reasonable?

**Comment by henrikbrixandersen** [comment]
> I propose that we test this from the `west flash` test cases once that tool understands how to react to these Kconfig options. Sound reasonable?

Yup.

**Comment by mbolivar-nordic** [comment]
You know, since this feature depends on west, we could always make relative paths relative to the workspace root. I won't do that for this PR, but I'll give it a shot when modifying `west flash` to use this.

**Comment by JordanYates** [comment]
Looks good to me

**Review by pabigot — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Just a rebase.

**Comment by mbolivar-nordic** [comment]
I added DNM because I would like to see https://github.com/zephyrproject-rtos/zephyr/pull/27740 merged first, so I can rebase on top without the extra shell out to west list.

File: cmake/mcuboot.cmake
Comment (tejlmand) [inline cmake/mcuboot.cmake:45]: I know this will change now, but remember, as per: https://github.com/zephyrproject-rtos/zephyr/pull/26060, we should avoid calling `west` directly, but instead through same Python as rest of build system.

File: cmake/mcuboot.cmake
Comment (tejlmand) [inline cmake/mcuboot.cmake:65]: This should be using same python as rest of build system instead.
See https://github.com/zephyrproject-rtos/zephyr/pull/26060

**Review by tejlmand — CHANGES_REQUESTED** [review]
We should use same python as rest of build system, when calling `west`.
See https://github.com/zephyrproject-rtos/zephyr/pull/26060.

Would `python -m west sign ...` be of help, or is that too early, cause it would force everyone to use a newer `west`

File: cmake/mcuboot.cmake
Comment (mbolivar-nordic) [reply cmake/mcuboot.cmake:65]: OK, I will need to think about how to do that in a convenient way.

I don't want to release 0.7.4 and would rather just get 0.8 out the door with this. We can add a hack for 0.7.x that breaks abstraction barriers. Cutting a new west release requires enough testing that I don't want to do it unless I have to :smile:

**Comment by mbolivar-nordic** [comment]
Open issues:

- now blocked on #27740 
- need to run west using the same python as everything else
- I've tested making relative paths relative to `west topdir` and it's working fine, so I'll push that into this PR since it was a trivial change and we're blocked now

I have a draft PR open for making `west flash` use the signed artifacts as well: https://github.com/zephyrproject-rtos/zephyr/pull/27818. It's not ready to merge, but I thought I'd mention it since this is back to being blocked in case people are curious.

**Comment by mbolivar-nordic** [comment]
Pushed a v4:

- rebased onto #27740 plus one DNM fix that ought to go into that PR (@tejlmand : I requested changes there)
- added a new patch, `cmake: west detection fixes and improvements`, that makes it so WEST runs west under the same python interpreter as the rest of the system (@tejlmand this addresses your open comment, please revisit)
- relative paths are now relative to `west topdir`

Open issues:

- blocked on #27740

File: cmake/west.cmake
Comment (tejlmand) [inline cmake/west.cmake:79]: How about only special handling of `0.7.x` like:

```suggestion
set(WEST_MODULE west)
if(${west_version} VERSION_LESS 0.8)
  # In west 0.7.x, west is not directly available as a module so we need to specify full namespace to main.
  string(APPEND WEST_MODULE ".app.main")
endif()
```

Another thing, a clever man once said:
> Same thing here; no need to expose the CLI implementation details. We can just use the west API directly.

(sorry, couldn't resist :laughing: )

**Review by tejlmand — COMMENTED** [review]
Looks good. Just a small suggestion.

File: cmake/west.cmake
Comment (mbolivar-nordic) [reply cmake/west.cmake]: > How about only special handling of `0.7.x` like:

Sure, that's fine.

Done and done.



> Another thing, a clever man once said:

Heh

**Comment by mbolivar-nordic** [comment]
Pushed v5:

- rebased on master now that #27740 is merged
- applied @tejlmand's suggested change to the new patch from v4

No open issues. @tejlmand please revisit and approve if satisfied.

**Review by tejlmand — APPROVED** [review]
lgtm

File: scripts/west_commands/sign.py
Comment (marc-hb) [inline scripts/west_commands/sign.py:323]: I really doubt this works. I also tried to https://pylint.pycqa.org/en/latest/user_guide/messages/warning/logging-not-lazy.html  somewhere else but it did not work; this is not the standard python logger unfortunately.

Comment (mbolivar-nordic) [reply scripts/west_commands/sign.py:323]: > I really doubt this works.

This PR was merged over 2 years ago; please be more specific or file a bug :)

Comment (marc-hb) [reply scripts/west_commands/sign.py:323]: Sorry for the noise, I didn't expect `log.dbg()` to work like this and notably to take a `level` argument while the other... levels (!) don't take any. It's not symmetric and intuitive.

What I should definitely have expected from you is for all that to be thoroughly documented:
https://docs.zephyrproject.org/1.14.0/guides/west/west-apis.html#module-west.log

My bad!

**Pull request merged**
